content/6_Advanced/Game_Theory.mdx

```diff
@@ -40,10 +38,389 @@ frequency: 0
+
+Suppose that $p$ is the position of the most significant bit in $y$.
+Then, we know that there must exist some $x_i$ with a set bit at $p$ as well.
+Whether $p$ is the most significant bit in $x_i$ or not, XOR-ing the two gives
+a result with a $0$ at bit $p$.
+This then guarantees that there exists an $x_i > y \oplus x_i$.
+
+<center>
+
+| Variable | Value |
+|------|------|
+| $y$    | `0...01?...?` |
+| $x_i$ | `?...?1?...?` |
+| $y \oplus x_i$ | `?...?0?...?` |
+
+*Since the first part of $y$ is all `0`s, the first parts of $x_i$ and $y \oplus x_i$ are the same.*
+
+</center>
+
+### Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+int main() {
+	int test_num;
+	std::cin >> test_num;
+	for (int t = 0; t < test_num; t++) {
+		int pile_num;
+		std::cin >> pile_num;
+		vector<int> piles(pile_num);
+		for (int &p : piles) { std::cin >> p; }
+
+		int xor_sum = 0;
+		for (int p : piles) { xor_sum ^= p; }
+
+		cout << (xor_sum != 0 ? "first" : "second") << '\n';
+	}
+}
+```
+
+</CPPSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Main {
+	public static void main(String[] args) throws IOException {
+		BufferedReader read =
+		    new BufferedReader(new InputStreamReader(System.in));
+		int testNum = Integer.parseInt(read.readLine());
+		for (int t = 0; t < testNum; t++) {
+			int pileNum = Integer.parseInt(read.readLine());
+			StringTokenizer pileST = new StringTokenizer(read.readLine());
+			int[] piles = new int[pileNum];
+			for (int p = 0; p < pileNum; p++) {
+				piles[p] = Integer.parseInt(pileST.nextToken());
+			}
+
+			int xorSum = 0;
+			for (int p : piles) { xorSum ^= p; }
+
+			System.out.println(xorSum != 0 ? "first" : "second");
+		}
+	}
+}
+```
+
+</JavaSection>
+<PySection>
+
+```py
+for _ in range(int(input())):
+	pile_num = int(input())
+	sizes = [int(p) for p in input().split()]
+	assert pile_num == len(sizes)
+
+	xor_sum = 0
+	for s in sizes:
+		xor_sum ^= s
+
+	print("first" if xor_sum != 0 else "second")
+```
+
+</PySection>
+</LanguageSection>
+
+<Optional title="What if we could add?">
+
+In some cases, players may be allowed to add sticks as well as remove them.
+Funnily enough, this doesn't change how the winning and losing states are determined,
+i.e. the outcome of the game.
+
+</Optional>
+
+## Sprague-Grundy theorem
+
+<Resources>
+	<Resource
+		source="CP Algorithms"
+		title="Sprague-Grundy theorem. NIM"
+		url="https://cp-algorithms.com/game_theory/sprague-grundy-nim.html"	/>
+	<Resource
+		source="Youtube"
+		title="Sprague-Grundy-Theorem"
+		url="https://www.youtube.com/watch?v=GRlGknQEOW8"
+		starred
+	/>
+</Resources>
+
+The [Sprague-Grundy theorem](https://en.wikipedia.org/wiki/Sprague%E2%80%93Grundy_theorem) defines "nimbers" or "grundy values" as
+
+$$
+\tag{1}
+n(x) = \text{mex}\{\text{all reachable nimber values}\}
```

Comment:
this is a recursive definition of what nimbers are

Reply:
It's not recursive. It's just what contains.

Reply:
you haven't explained what a nimber is

Reply:
`This means that every state of the game can be assigned a nimber value.`

I don't think it has a straight definition.

Reply:
wdym?

Reply:
Ok then it's a recursive def

Reply:
i mean do you see how this is a problem

Reply:
`defines "nimbers" or "grundy values" of any impartial game as the unique one-pile NIM game that the state of the game is equivalent to. These values can be computed using the recursive formula:`

I added this